### ISSUE TITLE
1.5.0.0 questions

### ISSUE THREAD

**Issue by joeyhub** [jimradford/superputty#543]
I've already started locally implementing a lot of changes that would make sense for the 1.5 milestone.

I would like to know if any other work is going on in the same direction to avoid difficulties merging in future and to try to coordinate if possible.

My biggest weaknesses will be lack of time to make tests, limited resources to test everything manually and some poor practices/rush class structure (I only just learn that C# has properties and fields as separate types of object variable members).

A key problem I am tackling is with the handling of large numbers of sessions and connections along with the data model representation of them.

The changes I am have made are the following:
* SessionData is in the Data namespace and more separated than before.
* SessionData now has a proper graph structure so that both folders and sessions are saved.
* Several small improvements and optimisations.
* Several performance improvements.

This has introduced some bugs because of the sheer number of changes which I am working on fixing (primarily the layout now and scp/file transfer features which I don't normally use).  I have also only spent a few days using C# so my code structure/standards will have some issues and a bit of inconsistency.

In the changes I have made so far the graph structure is far more convenient and adaptable. I am finding that it performs adequately now with over 10000 sessions in the list and has much less flickering for various small operations such as moving a node.

Creating a model system like this has brought about some immediate advantages. It is now possible to have any name for a node including repeat names or names with / in them. Folders are now also searchable and empty folders can be saved. It performs adequately with over 10000 sessions in a test and does not need to flicker or refresh the entire structure for small operations such as renaming or editing a node. Insert order/user order can also be preserved with this system.

However I made it like this mainly to make it adaptable to other data sources in the future with features such as special folder types that can receive data from another source.

Features I will add once stable:
* A network folder that is has a read only property for the treeview and a remote unique id property. This takes data from a streaming TCP service. This will be my own personal thing but will also to test the feasibility and help provide an initial pattern.
* A simple system for configuring commands that can be run for a host with right click templated from the properties in the session. In this iteration those windows will not be managed by superputty. I may only have time to implement this in the settings without a UI for configuring it.

Other bugs/issues:
* Hotkeys not registering when the putty window is focused. This is problematic for quick switching.
* Putty can be expensive for the CPU/GUI when it has a lot of activity. It should be confirmed whether or not putty can reduce load when out of view and if so if it does this with superputty.
* The full path text doesn't fit into the open sessions dialog and it is not resizable.
* In some situations the putty window steals focus when it shouldn't.
* DataView for open session crashes with a large number of sessions due to stack overflow.
* Sessions crashes if you close it and open it.
* Drop down menu for open tabs gets too big.

Other possible future features:
* Make kitty cmd work properly with ssh exec. ~ It's indirect.
* Connect N times. + Should be easy.
* Auto number prefix tabs to give simple unique names or mass rename with dataview. ~ Small enhancement.
* Ability to export config files for other apps, perhaps on a per folder basis. ~ Might be better as a personal plugin.
* Encrypted config with single password. +
* Manage pageant/xming. - A bat script can manage.
* Replace pageant with single password solution. ~ Depends how easy the pageant protocol is to copy.
* Copy a whole folder in the sessions view. +
* Allow the sessions list to be shown (cloned) as many times as the user likes so that different filters can be used. ~ It is not essential but could be easy.
* More advanced send commands filters using switch session like filter. ~ Also not essential. This kind of thing should be scripted.
* WPF? - Wont improve much.
* Tab groups? - A lot of work, small improvement.
* Application panel groups? - A lot of work, small improvement.
* In built SSH capability. - A lot of work, small improvement.I've already started locally implementing a lot of changes that would make sense for the 1.5 milestone.

I would like to know if any other work is going on in the same direction to avoid difficulties merging in future.

I expect to have a release candidate branch I can submit the Monday after next which will change around a couple dozen files.

The changes I am making this week or have made are the following:
* SessionData is in the Data namespace and more separated than before.
* SessionData now has a proper graph structure so that both folders and sessions are saved.
* Several small improvements and optimisations.
* Several performance improvements.

This has introduced some bugs because of the sheer number of changes which I am working on fixing.  I have also only spent a few days using C# so my code structure/standards will have some issues and a bit of inconsistency.

In the changes I have made so far the graph structure is far more convenient and adaptable. I am finding that it performs adequately now with over 10000 sessions in the list and has much less flickering for various small operations such as moving a node.

Creating a model system like this has brought about some immediate advantages. It is now possible to have any name for a node including repeat names or names with / in them. Folders are now also searchable and empty folders can be saved. It performs adequately now with over 10000 sessions in a test and does not need to flicker or refresh the entire structure for small operations such as renaming or editing a node. Insert order/user order can also be preserved with this system.

However I made it like this mainly to make it adaptable to other data sources in the future with features such as special folder types that can receive data from another source.

Next week:
* A network folder that is has a read only property for the treeview and a remote unique id property. This takes data from a streaming TCP service. This will be my own personal thing to test the feasibility.
* A simple system for configuring commands that can be run for a host with right click templated from the properties in the session. In this iteration those windows will not be managed by superputty. I may only have time to implement this in the settings without a UI for configuring it.

Other bugs/issues:
* Hotkeys not registering when the putty window is focused. This is problematic for quick switching.
* Putty can be expensive for the CPU/GUI when it has a lot of activity. It should be confirmed whether or not putty can reduce load when out of view and if so if it does this with superputty.
* The full path text doesn't fit into the open sessions dialog and it is not resizable.
* In some situations the putty window steals focus when it shouldn't.
* DataView for open session crashes with a large number of sessions due to stack overflow.
* Sessions crashes if you close it and open it.
* Drop down menu for open tabs gets too big.

Other possible future features that I am thinking about:
* Connect N times. + Should be easy.
* Auto number prefix tabs to give simple unique names or mass rename with dataview. ~ Small enhancement.
* Ability to export config files for other apps, perhaps on a per folder basis. ~ Might be better as a personal plugin.
* Encrypted config with single password. +
* Manage pageant/etc. - A bat script can manage.
* Replace pageant with single password solution. ~ Depends how easy the pageant protocol is to copy.
* Copy a whole folder in the sessions view. + It should be easy now and it makes sense.
* Allow the sessions list to be shown (cloned) as many times as the user likes so that different filters can be used. ~ It is not essential but could be easy.
* More advanced send commands filters using switch session like filter. ~ Also not essential. This kind of thing should be scripted.
* WPF? - Wont improve much.
* Tab groups? - A lot of work, small improvement.
* Application panel groups? - A lot of work, small improvement.
* In built SSH capability. - A lot of work, small improvement, also replacing scripts.

**Comment by waukeerules** [comment]
Sounds like a lot of improvements. I'd be happy to test new features and provide feedback.

**Comment by joeyhub** [comment]
I can't guarantee that I can make all of those improvements as that depends on my free time, however I hope to be able to submit my improved version soon. Locally it appears to be stable for all of my use cases.

**Comment by jimradford** [comment]
What is the purpose of encrypting the session/config files? Reason I bring it up is there is absolutely no secure way to store passwords. Even if they are encrypted in a file they will still show up in the process list since putty is executed with those command line arguments visible. I strongly encourage people who want to login with no passwords to use host/key authentication that is built into putty itself, or with pageant.
 
With that said, a more secure way to do so would be a login type script ala expect/answer where the password is being sent to the terminal over the ssh connection itself and not via command line arguments.

**Comment by jimradford** [comment]
Also, make sure any pull requests you submit are as specific to a single issue you are working on or a feature you are implementing as possible, this will minimize merge conflicts or at least make them easier to manage and backout if something goes terribly awry.

Thanks for your efforts!

**Comment by joeyhub** [comment]
Those features in the bottom list mostly are for consideration in the future and wont be applied to this release. I'm not 100% sure if I will implement any of those myself or not due to freetime. However putting it there lets others share their interests and thoughts as well as putting the ideas on the table. The only ones in that list done were easy/unobtrusive and were to avoid more branches than I could maintain.

I'm focused on the issues with the session data representation and various bugs rather than features with this update.

For the encryption that is only a proposed feature. However, on the session file encryption (password protected), even if the password shows in the processlist it does add more security. If someone wanted to put their putty sessions on portable media or a laptop, transfer it over the internet, etc; then this could be helpful. There are other ways to do that though not as convenient. Not all systems support public keys and must use password authentication. If someone is confident that their PC is secure it is not really a problem. However it is definitely something to think about with a lot of diagnostic tools or other otherwise innocent tools listing running processes that might scrape it up by accident. I was not thinking of encryption specifically for the args however. I had in mind for if features were added to manage/replace pageant and passphrases, etc and that the session data might be sensitive even without passwords. A lot of people are using superputty like a wallet for all their credentials. Perhaps it might be useful to put a simple check for -pw in the extra args with the edit dialog to popup a friendly warning to the user that they might be doing something insecure and link to pageant/explanation. To solve the problem fully kitty/putty would need to be modified. Expect is unreliable so ssh's explicit mechanisms for authentication should be used where available. It is good to have only for scenarios where there is no explicit authentication procedure.

The changes I have made are quite large because they fix a problem in the underlying structure of the software. I am keeping it uptodate with master so that it can integrate without conflicts. I rebased the first time without thinking but will merge in future. I've pushed to a candidate branch for early review.

https://github.com/joeyhub/superputty/tree/joeyhub_candidate_1.5

The are lots of little quality issues in that around quality (no commented/documented yet, some funny naming conventions, funny use of fields and properties, etc). If you have any feedback on conventions/standards you would like to see used let me know :).

**Comment by waukeerules** [comment]
I have a thought. How do you guys feel about creating a recent sessions menu?

**Comment by jimradford** [comment]
@joeyhub  I created a wiki page https://github.com/jimradford/superputty/wiki/Roadmap which I would encourage you to add your plans to. Seems like it would be a good idea for anyone adding code to have 1 place that has an overview of what we're working on (or maybe someone will see an itch and scratch it).

As far as standards go, hehe yes ultimately I would like for us to all follow some. I generally like the ones published by Microsoft and try to adhere to them. Sadly much refactoring will be required to get us 'back' to that state.

Basically: Compiles with no warnings (or errors) are GOOD
Source code comments are GOOD, but it is not necessary to overdo it. All public functions, etc should be described and have their parameters documented, private and internal ones it would be nice, and maybe we'll get there as well, I don't have huge expectations here.

**Comment by jimradford** [comment]
@waukeerules I don't feel bad about it. Could you open a new issue for that request?

**Comment by joeyhub** [comment]
I have added to that and cleaned up a bit.

I think the only issue on this topic is that we are both tackling the same issue with other sources for sessions. However, I don't think our approaches are incompatible with each other. It looks like one is bottom up and one is top down.

**Comment by jimradford** [comment]
@joeyhub Thanks. I had previously implemented a plugin system as a proof of concept for managing sessions, I created an Interface class to abstract the management of sessions outside of the core application, I had implemented a (default) SaveToDisk plugin and a SaveToWeb, the SaveToDisk plugin simple refactored the current session save and load routines to an external plugin, the SaveToWeb was to be used as an example of how to use the Plugin system (I think I had hacked together a php script for the webserver). I got sidetracked by fixing bugs and scratching other itches in the application that arose. I am not tied to that implementation, if I were It would be a rewrite of my proof of concept. One thing we might consider is MEF. When MEF first came out I got all hot and bothered by it and built a pretty fun plugin system for another project. Of course that was years ago and I've long since forgotten more than I remember. At this point I don't see where plugins make sense for other parts of the application, But for session data it certainly does.

Other than bug fixing I don't have big plans for any major changes now that 1.4.0.7 is released, especially being mindful of the work you're doing right now. So I look forward to seeing what you come up with.

**Comment by joeyhub** [comment]
The MEF looks like an interesting concept. I think if it is made to work with plugins it would be easier to focus on the data conversion layer with some options passed up for the UI such as readonly, etc rather than providing hooks in the first stage. I am experimenting with that so later we could have a better idea of what would be required.

I'm having one set of conundrums in particular related to where to put the sources. If you take a look at problematic issues and the one about circular references close to the bottom your thoughts on that would be most appreciated:
https://github.com/joeyhub/superputty/wiki/Notes

My branch should be buildable and hopefully possible to play around with if it doesn't crash on trying to load legacy settings (all settings should be backed up first and restored after otherwise it might break them):
https://github.com/joeyhub/superputty/tree/joeyhub_candidate_1.5

**Comment by jimradford** [comment]
I am in Prague for the next couple weeks, but will look into this more when I arrive back home.

**Comment by bamanzi** [comment]
> A simple system for configuring commands that can be run for a host with right click templated from Host in the session. So far only had time to implement this in the settings file without a UI for configuring it.

I'm very looking forward to this feature. Sometimes I want to launch a WinSCP session or a MobaXterm session from the tree.  A few weeks ago I was using mRemoteNG, which is good at this (e.g. [3 Tips for the mRemoteNG Remote Connections Manager](http://www.howtogeek.com/175692/3-tips-for-the-mremoteng-remote-connections-manager/ ).
